### PR TITLE
Game logic improvements

### PR DESCRIPTION
**Pull request by JereHirviniemi**

### CODE REVIEW

**Comment by JereHirviniemi** [comment]
@ketoni pls fix states

**Comment by ketoni** [comment]
One more rebase to origin/pvtrivia to get @JonneOkkonen's fix to question loading

File: Project P/PVTrivia/States/MainState.cs [hunk 1]
```diff
@@ -32,28 +40,100 @@ public override void Update(float timeStep)
...
+            //Create 10 floors visible immediately
+            int i = 1;
+            while (i <= 10)
+            {
+                CreateFloor(i, SceneRoot.CreateChild("Floor" + i));
+                i++;
+            }
 
-            jumpV = new Vector2(player.GetComponent<RigidBody2D>().LinearVelocity.X, 10);
+            World.PhysicsBeginContact2D += (PhysicsBeginContact2DEventArgs obj) =>
```
Comment (ketoni) [inline Project P/PVTrivia/States/MainState.cs]: This causes multiple actions to fire when there are 2+ `MainState`s created, so remove the old action in `State.Exit()`.
This should fix the crash mentioned in https://trello.com/c/lbrk3SXK

File: Project P/PVTrivia/States/MainState.cs [hunk 2]
```diff
@@ -66,65 +98,149 @@ public override void Resume()
...
+        {
+            if (player.Position.Y > bestHeight - 13)
+            {
+                pBody.SetLinearVelocity(new Vector2(args.DX / 2, pBody.LinearVelocity.Y));
+            }
+        }
+
+        public override Node Exit()
+        {
+            Context.Input.TouchMove -= Movement;
```
Comment (JonneOkkonen) [inline Project P/PVTrivia/States/MainState.cs]: Add ChangeState to Exit

`PVTrivia.States.ChangeState(new EndState(9000));`

File: Project P/PVTrivia/States/MainState.cs [hunk 3]
```diff
@@ -66,65 +98,149 @@ public override void Resume()
...
+        private void Movement(TouchMoveEventArgs args)
+        {
+            if (player.Position.Y > bestHeight - 13)
+            {
+                pBody.SetLinearVelocity(new Vector2(args.DX / 2, pBody.LinearVelocity.Y));
+            }
+        }
+
+        public override Node Exit()
+        {
```
Comment (JonneOkkonen) [inline Project P/PVTrivia/States/MainState.cs]: Add ChangeState to Exit

`PVTrivia.States.ChangeState(new EndState(9000));`

File: Project P/PVTrivia/States/MainState.cs [hunk 4]
```diff
@@ -66,65 +98,149 @@ public override void Resume()
...
-            //Sets the sprite
-            StaticSprite2D spriteComponent = parentNode.CreateComponent<StaticSprite2D>();
-            spriteComponent.Sprite = cache.GetSprite2D("Textures/Player.png");
+            /*
+            //Sets the triggerbox for checking whether player is inside another collision box
+            Node triggerNode = parentNode.CreateChild("TriggerNode");
+            CollisionBox2D triggerbox = triggerNode.CreateComponent<CollisionBox2D>();
+            triggerbox.Trigger = true;
+            triggerbox.SetSize(playerW, playerH);
+            */
```
Comment (ketoni) [inline Project P/PVTrivia/States/MainState.cs]: Is this needed?

File: Project P/PVTrivia/States/MainState.cs [hunk 5]
```diff
@@ -32,28 +43,49 @@ public override void Update(float timeStep)
...
             {
-                if (obj.NodeA.Name == "Player") obj.BodyA.SetLinearVelocity(jumpV);
-                PVTrivia.Log("NodeA = " + obj.NodeA.Name);      //temp
-                PVTrivia.Log("NodeB = " + obj.NodeB.Name);      //temp
-            };
+                pHitbox.Enabled = true;
+            }
+
+            //Ends the game if the player falls 1500px from their bestHeight
+            if (player.Position.Y < bestHeight-15) End();
```
Comment (ketoni) [inline Project P/PVTrivia/States/MainState.cs]: Using brackets is encouraged! 😨 
Please add them here.

File: Project P/PVTrivia/States/MainState.cs (hunk 1 above)
Comment (JereHirviniemi) [reply Project P/PVTrivia/States/MainState.cs]: Done

**Pull request merged**